### PR TITLE
[Feature] Filter Criteria for Late Filtering

### PR DESCRIPTION
**Pull request by net-cscience-raphael**
# PR Summary

This PR addresses **#121** and introduces a query transformer, `LateFilter`, which enables filtering `retrievables` at a later stage in the query processing pipeline.

Similar to the `FieldLookup` transformer, `LateFilter` fetches a specified `field`. `LateFilter` then compares its `value` against a predicate. Only `retrievables` that satisfy the predicate are emitted.

## Setup and Usage

The `LateFilter` can be configured by adding the transformer to the query, as shown in the following example:

```json
{
  "inputs": {
    "input-text": {"type": "TEXT", "data": "an orange starfish on the seafloor"}
  },
  "operations": {  
   "clip" : {"type": "RETRIEVER", "field": "clip", "input": "input-text"},
   "relations" : {"type": "TRANSFORMER", "transformerName": "RelationExpander", "input": "clip"},
   "aggregator" : {"type": "TRANSFORMER", "transformerName": "ScoreAggregator",  "input": "relations"},
  "lookup0" : {"type": "TRANSFORMER", "transformerName": "FieldLookup", "input": "aggregator"},
  "lookup1" : {"type": "TRANSFORMER", "transformerName": "FieldLookup", "input": "lookup0"},
   "filter0" : {"type": "TRANSFORMER", "transformerName": "LateFilter", "input": "lookup1"},
  "filter1" : {"type": "TRANSFORMER", "transformerName": "LateFilter", "input": "filter0"},
  "filter2" : {"type": "TRANSFORMER", "transformerName": "LateFilter", "input": "filter1"}
  },
  "context": {
   "global": {    
   "limit": "70000"},
   "local" : {    
    "relations" :{"outgoing": "partOf"},    
    "lookup0": {"field": "whisperasr", "keys": "value"},
    "lookup1": {"field": "time", "keys": "start, end"},
    "filter0": {"field": "whisperasr", "keys": "value",  "comparison": "~=", "value": "%", "skip": "ignore"},
    "filter1": {"field": "time", "keys": "start",  "comparison": ">=", "value": "1", "skip": "ignore"},
    "filter2": {"field": "time", "keys": "end",  "comparison": "<=", "value": "3500000000000", "skip": "ignore", "limit": "100"}
   }
  },
  "output": "filter2"
}
```
The `FieldLookupLateFilter` can be configured by adding the transformer to the query, as shown in the following example:
```json
{
  "inputs": {
    "input-text": {"type": "TEXT", "data": "an orange starfish on the seafloor"}
  },
  "operations": {  
   "clip" : {"type": "RETRIEVER", "field": "clip", "input": "input-text"},
   "relations" : {"type": "TRANSFORMER", "transformerName": "RelationExpander", "input": "clip"},
   "aggregator" : {"type": "TRANSFORMER", "transformerName": "ScoreAggregator",  "input": "relations"},
   "filter0" : {"type": "TRANSFORMER", "transformerName": "FieldLookupLateFilter", "input": "aggregator"},
  "filter1" : {"type": "TRANSFORMER", "transformerName": "FieldLookupLateFilter", "input": "filter0"},
  "filter2" : {"type": "TRANSFORMER", "transformerName": "FieldLookupLateFilter", "input": "filter1"}
  },
  "context": {
   "global": {  
 "limit": "70000"},
   "local" : {    
    "relations" :{"outgoing": "partOf"},    
  "lookup0": {"field": "whisperasr", "keys": "value"},
  "lookup1": {"field": "time", "keys": "start, end"},
  "filter0": {"field": "whisperasr", "keys": "value",  "comparison": "~=", "value": "%", "append": "true"},
  "filter1": {"field": "time", "keys": "start",  "comparison": ">=", "value": "1", "append": "true"},
  "filter2": {"field": "time", "keys": "end",  "comparison": "<=", "value": "3500000000000", "append": "true"}
   }
  },
  "output": "filter2"
}
```

## Supported Comparison Operations

The LateFilter supports the following comparison operations:

- `==` : Equals
- `!=` : Not equal
- `<` : Less than
- `>` : Greater than
- `<=` : Less than or equal to
- `>=` : Greater than or equal to
- `~=` : Like (pattern matching)

## Additional Configuration Options

`LateFilter` provides the following options:
- `"append": Boolean`        If set to true, the specified property is added to the retrievables (similar to FieldLookup).
- `"limit": Int` Limits the number of emitted retrievables to the specified value.

### CODE REVIEW

**Comment by lucaro** [comment]
I'm not sure this is the best approach. I would either have an operator that pushes down the filtering to the database or one that filters based on existing information, such as generated by the `FieldLookup`.

**Comment by ppanopticon** [comment]
> I'm not sure this is the best approach. I would either have an operator that pushes down the filtering to the database or one that filters based on existing information, such as generated by the `FieldLookup`.

It certainly isn't the best approach, but that's not the goal here.

- Your first suggestion is - of course - the desired outcome. But the facilities in place are not geared towards this yet and making this happen (end-to-end) constitutes a major undertaking.
- The latter may be better in terms of code repetition. But if we think in terms of execution planning (i.e., thinking towards the ideal end goal), spreading such a primitive over multiple operations will complicate the planning process.

**Comment by lucaro** [comment]
Then at the very least, the logic should check if the field is already present before again performing a lookup.

File: vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/model/query/basics/Distance.kt [hunk 1]
```diff
@@ -98,6 +98,19 @@ enum class Distance {
         override fun invoke(v1: Value.DoubleVector, v2: Value.DoubleVector): Double = throw UnsupportedOperationException("Jaccard distance is not supported for float vectors.")
     };
 
+    companion object {
+        infix fun fromString(value: String): Distance {
```
Comment (lucaro) [inline vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/model/query/basics/Distance.kt:102]: What is wrong with `Distance.fromString(value.toUpper())` ?

**Pull request merged**